worker/worker.go

```diff
@@ -495,29 +507,64 @@ func (w *worker) rhpFundHandler(jc jape.Context) {
 	if jc.Decode(&rfr) != nil {
 		return
 	}
-	rk := w.deriveRenterKey(rfr.HostKey)
-	err := w.withTransportV3(jc.Request.Context(), rfr.HostIP, rfr.HostKey, func(t *rhpv3.Transport) (err error) {
-		// The FundAccount RPC requires a SettingsID, which we also have to pay
-		// for. To simplify things, we pay for the SettingsID using the full
-		// amount, with the "refund" going to the desired account; we then top
-		// up the account to cover the cost of the two RPCs.
-		payment, ok := rhpv3.PayByContract(&rfr.Contract, rfr.Amount, rfr.Account, rk)
-		if !ok {
-			return errors.New("insufficient funds")
-		}
-		priceTable, err := rhpv3.RPCPriceTable(t, &payment)
+	// Get account for the host.
+	account, err := w.accounts.AccountForHost(rfr.HostKey)
+	if jc.Check("failed to get account for provided host", err) != nil {
+		return
+	}
+
+	// Get IP of host.
+	h, err := w.bus.Host(rfr.HostKey)
+	if jc.Check("failed to fetch host", err) != nil {
+		return
+	}
+	hostIP := h.Settings.SiamuxAddr()
+
+	// Get contract revision.
+	// TODO: This is not a good solution. We should have some mechanism to
+	// lock contracts across protocol versions. This is abusing rhpv2's
+	// withHosts to get the revision.
+	var revision types.FileContractRevision
```

Comment:
I'd appreciate some input on this. I think this is going to bite us during times of heavy uploads/migrations. We need some way to lock a contract across renter-host protocols for things that require revising a contract.

Reply:
This is very tricky indeed. It would have to live in the `bus` right and the workers would have to respect the locks for anything when they're not paying for it using an ephemeral account. I'm not sure if that's feasible. Crazy idea: what if we keep track of revisions and use them optimistically. If it fails we lock the contract but if it succeeds we win? Haven't given it much thought, happy to discuss this on discord tomorrow. It is really tricky I don't immediately have a good solution or suggestion.

Reply:
Hosts return the price table regardless of payment right, we could ab(use) that for the time being. I'll think about it some more.

Reply:
Yeah, so the problem disappears once everything that revises contracts is triggered by the autopilot and anything else simply uses EAs. Because then the autopilot can just lock contracts in memory.

Ideally we only fetch a price table with a contract once anyway. After that it should be done through EAs only.
So let's keep an eye on it for now and see if it becomes an issue.